### DISCUSSION TITLE
Batch initializing availability database objects

### DISCUSSION THREAD

**Discussion by aguzev** [microsoft/sqlmanagementobjects#109] [Q&A] [answered]
The task is to enumerate all availability databases in all availability groups on a server. SMO executes a query for each availability database in each AG. The number of trips to the SQL Server grows with the number of databases.

Is it possible to configure SMO to get the data as a single batch with one or a couple of queries?

**Comment by shueybubbles** [comment, marked as answer]
@aguzev what does the naive code loop look like?

**Reply by aguzev** [reply]
The pseudocode:

```CSharp
foreach(var ag in server.AvailabilityGroups)
    foreach(var state in ag.DatabaseReplicaStates)
    {
        // do something…
    }
```

We've found the following solution. Is it the right one?

```CSharp
var queryObj = new SfcObjectQuery(server);
var sfcQuery = new SfcQueryExpression("Server/AvailabilityGroup/DatabaseReplicaState");
foreach(var state in queryObj.ExecuteIterator(sfcQuery, fields, null))
{
    // do something…
}
```

**Reply by shueybubbles** [reply]
it looks reasonable to me, though I think you can use a `Request` object and `server.ExecutionManager.GetEnumeratorData` to get a data table.
